### PR TITLE
[GORDO-1509] Use actor framework in pregel

### PR DESCRIPTION
**Pull request by jvolmer**
Uses actor framework in pregel:
When the pregel run now starts, a conductor actor is started. As soon as the involved DBServers are known, one worker actor is started on each of these servers. The conductor and worker actors are currently very simple. Code from the worker and conductor will be moved into these actors in following PRs.

For that to work, this PR:
- Adds an external dispatcher for arangodb and handles messages from this external dispatcher in the RestPregelHandler.
- To send REST messages, we need the specific database we are working on. Therefore, the database is now part of the ActorPID.
- To spawn an actor on another server, we start a spawn actor on each involved server. Because of the database addition to our ActorPID, the spawn server needs to be specific per database, therefore this actor can only be started once a pregel run started and the database is known.
Per REST this SpawnActor is created on the other server by sending a spawn message to this other server with the special Actor ID 0: The PregelRestHandler recognizes this special message by its ActorID 0 and spawns a SpawnActor.

### CODE REVIEW

File: arangod/Pregel/ArangoExternalDispatcher.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+#include "Actor/ActorPID.h"
+#include "Actor/Message.h"
+#include "CrashHandler/CrashHandler.h"
+#include "Assertions/ProdAssert.h"
+#include "Inspection/VPackWithErrorT.h"
+#include "Logger/LogMacros.h"
+#include "Network/ConnectionPool.h"
+#include "Network/Methods.h"
+
```
Comment (markuspf) [inline arangod/Pregel/ArangoExternalDispatcher.h:11]: There should be some documentation here about what the responsibility of this ArangoExternalDispatcher is

File: arangod/Pregel/ArangoExternalDispatcher.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    auto networkMessage =
+        NetworkMessage{.sender = sender,
+                       .receiver = receiver,
+                       .payload = velocypack::Builder(msg.slice())};
+    auto serialized = inspection::serializeWithErrorT(networkMessage);
+    ADB_PROD_ASSERT(serialized.ok());
+    auto builder = velocypack::Builder(serialized.get().slice());
+    return network::sendRequestRetry(
+        this->connectionPool,
+        // TODO: what about "shard:"?
```
Comment (markuspf) [inline arangod/Pregel/ArangoExternalDispatcher.h:59]: what about `shard:`?

File: arangod/Pregel/ArangoExternalDispatcher.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                                       out.errorMessage());
+                  }
+                });
+          }
+        });
+  }
+
+  network::ConnectionPool* connectionPool;
+  std::string baseURL;
+  network::Timeout timeout{5.0 * 60};
```
Comment (markuspf) [inline arangod/Pregel/ArangoExternalDispatcher.h:93]: magic constant; I am sure this comes from "legacy" code; is it worthwhile making this configurable?

File: arangod/Pregel/Conductor/Actor.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+template<typename Runtime>
+struct ConductorHandler : actor::HandlerBase<Runtime, ConductorState> {
+  auto operator()(ConductorStart start) -> std::unique_ptr<ConductorState> {
+    LOG_TOPIC("56db0", INFO, Logger::PREGEL)
+        << fmt::format("Conductor Actor {} started", this->self);
+    // TODO call State::initializeWorkers in here
+    // and then spawn actors from here
+    return std::move(this->state);
```
Comment (markuspf) [inline arangod/Pregel/Conductor/Actor.h:46]: when? Is this a future PR?

File: arangod/Pregel/PregelFeature.cpp [hunk 5]
```diff
@@ -274,15 +283,38 @@ ResultT<ExecutionNumber> PregelFeature::startExecution(TRI_vocbase_t& vocbase,
   }
 
   auto en = createExecutionNumber();
+
+  // TODO needs to be part of the conductor state
```
Comment (markuspf) [inline arangod/Pregel/PregelFeature.cpp:287]: Please create a ticket for this TODO

File: arangod/Pregel/SpawnActor.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        << "Spawn Actor: Spawn conductor actor";
+    this->template spawn<ConductorActor>(ConductorState{}, ConductorStart{});
+    return std::move(this->state);
+  }
+
+  auto operator()(SpawnWorker msg) -> std::unique_ptr<SpawnState> {
+    LOG_TOPIC("2452c", INFO, Logger::PREGEL)
+        << "Spawn Actor: Spawn worker actor";
+    this->template spawn<WorkerActor>(
+        WorkerState{}, WorkerStart{});  //.conductor = msg.conductor});
```
Comment (markuspf) [inline arangod/Pregel/SpawnActor.h:60]: why is this commented out?

File: arangod/Pregel/SpawnActor.h [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      -> std::unique_ptr<SpawnState> {
+    LOG_TOPIC("7b602", INFO, Logger::PREGEL) << fmt::format(
+        "Spawn Actor: Error - sent unknown message to {}", unknown.receiver);
+    return std::move(this->state);
+  }
+
+  auto operator()(actor::ActorNotFound notFound)
+      -> std::unique_ptr<SpawnState> {
+    LOG_TOPIC("03156", INFO, Logger::PREGEL) << fmt::format(
+        "Spawn Actor: Error - recieving actor {} not found", notFound.actor);
```
Comment (markuspf) [inline arangod/Pregel/SpawnActor.h:74]:
```suggestion
        "Spawn Actor: Error - receiving actor {} not found", notFound.actor);
```

File: arangod/Pregel/Worker/Messages.h [hunk 8]
```diff
@@ -24,13 +24,31 @@
...
+#include "Inspection/Types.h"
 #include "Pregel/ExecutionNumber.h"
 #include "Pregel/GraphStore/Graph.h"
 #include "Pregel/Statistics.h"
 #include "Pregel/Status/Status.h"
 
 namespace arangodb::pregel {
 
+struct WorkerStart {
+  // arangodb::pregel::actor::ActorPID conductor;
```
Comment (markuspf) [inline arangod/Pregel/Worker/Messages.h:36]: why is this commented out?

File: arangod/Pregel/Actor/include/Actor/ActorPID.h [hunk 9]
```diff
@@ -69,11 +69,14 @@ using DatabaseName = std::string;
 
 struct ActorPID {
   ServerID server;
+  DatabaseName database;
```
Comment (markuspf) [inline arangod/Pregel/Actor/include/Actor/ActorPID.h:72]: here is some architectural change in this PR that is not related to using Actors? 

In particular actors are now tied to databases?

File: tests/Pregel/Actor/ActorTest.cpp [hunk 10]
```diff
@@ -139,47 +140,45 @@ TYPED_TEST(ActorTest, does_not_work_on_new_messages_after_actor_finished) {
...
+  actor->finish();
+
+  // wait for actor to work off all messages
+  while (not actor->isIdle()) {
+  }
+  this->scheduler->stop();
+  ASSERT_EQ(actor->getState(),
+            (TrivialState{.state = std::string(sent_message_count, 'A'),
+                          .called = sent_message_count}));
+}
```
Comment (markuspf) [inline tests/Pregel/Actor/ActorTest.cpp:184]: Probably should be cautious that this re enables the unstable test (and I think this is my fault because I rebased on devel and reenabled it)

**Review by markuspf — COMMENTED** [review]
There are a few questions

File: arangod/Pregel/ArangoExternalDispatcher.h (hunk 2 above)
Comment (jvolmer) [reply arangod/Pregel/ArangoExternalDispatcher.h:59]: Yes, I'm not yet sure how to solve that. `shard:` is used (instead of `server:`) in OutgoingCache, here I think the receiving server is not known. In the actor framework we can currently address actors only via server id, so here we possibly need to change something.

File: arangod/Pregel/ArangoExternalDispatcher.h (hunk 3 above)
Comment (jvolmer) [reply arangod/Pregel/ArangoExternalDispatcher.h:93]: Yes, it is coming from the legacy code, you can now configure it in the constructor

File: arangod/Pregel/PregelFeature.cpp (hunk 5 above)
Comment (jvolmer) [reply arangod/Pregel/PregelFeature.cpp:287]: this will also be part of the next actor-related PR

File: arangod/Pregel/SpawnActor.h (hunk 6 above)
Comment (jvolmer) [reply arangod/Pregel/SpawnActor.h:60]: this will also be part of the next PR. Currently, the worker actors are spawned from Conductor.cpp. Next PR will move this part of Conductor.cpp into the conductor actor. Then the worker actor are spawned from within the conductor actor.

File: arangod/Pregel/Worker/Messages.h (hunk 8 above)
Comment (jvolmer) [reply arangod/Pregel/Worker/Messages.h:36]: See previous comment

File: arangod/Pregel/Actor/include/Actor/ActorPID.h (hunk 9 above)
Comment (jvolmer) [reply arangod/Pregel/Actor/include/Actor/ActorPID.h:72]: I think this is needed to use actors in ArangoDB: REST messages can only be send to specific databases, so the database name really needs to be part of the actor address, otherwise we cannot send messages to other servers. I tried to do it differently but failed, if you have a better idea I'm up for it.

**Review by markuspf — APPROVED** [review]
